File: iherefor-smart-development/demo/reader-app/android/app/src/main/java/com/iherefor/readerapp/data/LibraryViewModel.kt
package com.iherefor.readerapp.data

import com.iherefor.readerapp.domain.Book
import com.iherefor.readerapp.domain.BookFormat
import com.iherefor.readerapp.domain.ReadingProgress
import com.iherefor.readerapp.domain.SharedRules
import java.util.UUID

/** 书架用例逻辑（依赖注入仓库） */
class LibraryViewModel(private val repository: LibraryRepository) {

    fun books(): List<Book> = repository.books()

    /** 导入一本书：格式校验 + sourceUri 去重。返回错误消息则拒绝导入。 */
    fun importBook(ext: String, sourceUri: String, title: String): String? {
        val format = BookFormat.fromExtension(ext) ?: return "不支持的格式"
        if (SharedRules.findDuplicate(sourceUri, repository.books()) != null) {
            return "已存在"
        }
        val book = Book(
            id = UUID.randomUUID().toString(),
            title = title,
            format = format,
            sourceUri = sourceUri
        )
        repository.add(book)
        return null
    }

    fun removeBook(id: String) { repository.remove(id) }

    fun saveProgress(bookId: String, progress: ReadingProgress) {
        val book = repository.books().firstOrNull { it.id == bookId } ?: return
        if (SharedRules.isProgressCompatible(book.format, progress)) {
            repository.updateProgress(bookId, progress)
        }
    }
}

File: iherefor-smart-development/demo/reader-app/ios/ReaderApp/Data/LibraryViewModel.swift
import Foundation

/// 书架 ViewModel：承载导入、去重、删除等用例逻辑（依赖注入仓库）
final class LibraryViewModel: ObservableObject {

    private let repository: LibraryRepository

    init(repository: LibraryRepository) {
        self.repository = repository
    }

    func books() -> [Book] {
        repository.books()
    }

    /// 导入一本书：格式校验 + sourceUri 去重。返回错误则拒绝导入。
    @discardableResult
    func importBook(ext: String, sourceUri: String, title: String) -> String? {
        guard let format = SharedRules.detectFormat(fromExtension: ext) else {
            return "不支持的格式"
        }
        if let dup = SharedRules.findDuplicate(sourceUri: sourceUri, in: repository.books()) {
            return dup.id == "" ? nil : "已存在"  // 去重，不重复添加
        }
        let book = Book(id: UUID().uuidString, title: title, author: nil,
                        format: format, sourceUri: sourceUri, coverUri: nil,
                        importedAt: Date(), progress: nil)
        repository.add(book)
        return nil
    }

    func removeBook(id: String) {
        repository.remove(id: id)
    }

    func saveProgress(bookId: String, progress: ReadingProgress) {
        guard let book = repository.books().first(where: { $0.id == bookId }) else { return }
        guard SharedRules.isProgressCompatible(format: book.format, progress: progress) else { return }
        repository.updateProgress(bookId: bookId, progress: progress)
    }
}
